symengine/fields.cpp

```diff
@@ -0,0 +1,272 @@
+                            const Ptr<GaloisFieldDict> &quo,
+                            const Ptr<GaloisFieldDict> &rem) const
+{
+    map_uint_mpz dict_quo, dict_rem;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        if (iter.first >= n_val)
+            dict_quo.insert(dict_quo.end(), {iter.first - n_val, iter.second});
+        else
+            dict_rem.insert(dict_rem.end(), iter);
+    }
+    *quo = GaloisFieldDict(dict_quo, modulo_);
+    *rem = GaloisFieldDict(dict_rem, modulo_);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_sqr() const
+{
+    return (*this * *this);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_pow(const integer_class n) const
+{
+    if (n == 0) {
+        return GaloisFieldDict({integer_class(1)}, modulo_);
+    }
+    if (n == 1)
+        return static_cast<GaloisFieldDict>(*this);
+    if (n == 2)
+        return gf_sqr();
+    long num = mp_get_si(n);
+    GaloisFieldDict to_sq = static_cast<GaloisFieldDict>(*this);
+    GaloisFieldDict to_ret = GaloisFieldDict({integer_class(1)}, modulo_);
+    while (1) {
+        if (num & 1) {
+            to_ret *= to_sq;
+        }
+        num >>= 1;
+        if (num == 0)
+            return to_ret;
+        to_sq = to_sq.gf_sqr();
+    }
+}
+
+void GaloisFieldDict::gf_monic(integer_class &res,
+                           const Ptr<GaloisFieldDict> &monic) const
+{
+    if (dict_.empty()) {
+        res = integer_class(0);
+        *monic = static_cast<GaloisFieldDict>(*this);
+    } else {
+        res = dict_.rbegin()->second;
+        if (res == integer_class(1)) 
+            *monic = static_cast<GaloisFieldDict>(*this);
+        else {
+            monic->dict_.clear();
+            monic->modulo_ = modulo_;
+            integer_class inv, temp;
+            mp_invert(inv, res, modulo_);
+            for (auto &iter : dict_) {
+                mp_fdiv_r(temp, inv * iter.second, modulo_);
+                if (temp != integer_class(0))
+                    monic->dict_[iter.first] = temp;
+            }
+        }
+    }
+}
+
+GaloisFieldDict GaloisFieldDict::gf_gcd(const GaloisFieldDict &o) const
+{
+    SYMENGINE_ASSERT(modulo_ == o.modulo_);
+    GaloisFieldDict f = static_cast<GaloisFieldDict>(*this);
+    GaloisFieldDict g = o;
+    GaloisFieldDict temp_out, temp_ex;
+    while (not g.dict_.empty()) {
+        temp_ex = g;
+        f.gf_div(temp_ex, outArg(temp_out), outArg(g)); // g = f % g
+        f = temp_ex;
+    }
+    integer_class temp_LC;
+    f.gf_monic(temp_LC, outArg(temp_out));
+    return std::move(temp_out);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_lcm(const GaloisFieldDict &o) const
+{
+    SYMENGINE_ASSERT(modulo_ == o.modulo_);
+    if (dict_.empty())
+        return static_cast<GaloisFieldDict>(*this);
+    if (o.dict_.empty())
+        return o;
+    GaloisFieldDict out, temp_out;
+    out = o * (*this);
+    out /= gf_gcd(o);
+    integer_class temp_LC;
+    out.gf_monic(temp_LC, outArg(temp_out));
```

Comment:
If the changes in `gf_monic` are done, this can be written as `out.gf_monic(temp_LC, outArg(out));`